pmd-core/src/main/java/net/sourceforge/pmd/RuleBuilder.java

```diff
@@ -0,0 +1,246 @@
+ * @since 6.0.0
+ */
+class RuleBuilder {
+
+    private List<PropertyDescriptor<?>> definedProperties = new ArrayList<>();
+    private String name;
+    private String clazz;
+    private Language language;
+    private String minimumVersion;
+    private String maximumVersion;
+    private String since;
+    private String message;
+    private String rulesetName;
+    private String externalInfoUrl;
+    private String description;
+    private List<String> examples = new ArrayList<>(1);
+    private RulePriority priority;
+    private boolean isDeprecated;
+    private boolean isUsesDfa;
+    private boolean isUsesMultifile;
+    private boolean isUsesTyperesolution;
+
+
+    RuleBuilder(String name, String clazz, String language) {
+        this.name = name;
+        language(language);
+        className(clazz);
+    }
+
+
+    public void usesDFA(boolean usesDFA) {
+        isUsesDfa = usesDFA;
+    }
+
+
+    public void usesMultifile(boolean usesMultifile) {
+        isUsesMultifile = usesMultifile;
+    }
+
+
+    public void usesTyperesolution(boolean usesTyperesolution) {
+        isUsesTyperesolution = usesTyperesolution;
+    }
+
+
+    private void language(String languageName) {
+        if (StringUtils.isBlank(languageName)) {
+            return;
+            // throw new IllegalArgumentException("Blank language attribute");
+        }
+
+        Language lang = LanguageRegistry.findLanguageByTerseName(languageName);
+        if (lang == null) {
+            throw new IllegalArgumentException(
+                    "Unknown Language '" + languageName + "' for rule" + name + ", supported Languages are "
+                    + LanguageRegistry.commaSeparatedTerseNamesForLanguage(LanguageRegistry.findWithRuleSupport()));
+        }
+        language = lang;
+    }
+
+
+    private void className(String className) {
+        if (StringUtils.isBlank(className)) {
+            throw new IllegalArgumentException("The 'class' field of rule can't be null, nor empty.");
+        }
+
+        this.clazz = className;
+    }
+
+
+    public void minimumLanguageVersion(String minimum) {
+        minimumVersion = minimum;
+    }
+
+
+    public void maximumLanguageVersion(String maximum) {
+        maximumVersion = maximum;
+    }
+
+
+    private void checkLanguageVersionsAreOrdered(Rule rule) {
+        if (rule.getMinimumLanguageVersion() != null && rule.getMaximumLanguageVersion() != null
+            && rule.getMinimumLanguageVersion().compareTo(rule.getMaximumLanguageVersion()) > 0) {
+            throw new IllegalArgumentException(
+                    "The minimum Language Version '" + rule.getMinimumLanguageVersion().getTerseName()
+                    + "' must be prior to the maximum Language Version '"
+                    + rule.getMaximumLanguageVersion().getTerseName() + "' for Rule '" + name
+                    + "'; perhaps swap them around?");
+        }
+    }
+
+
+    public RuleBuilder since(String sinceStr) {
```

Comment:
why is this the only method to return the builder instance? Either we provide a fluent API or we don't

Reply:
Right